File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/annotation/AnnotationController.kt
package com.mapbox.maps.mapbox_maps.annotation

import com.mapbox.maps.MapView
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.mapbox_maps.pigeons.PointAnnotation
import com.mapbox.maps.plugin.annotation.Annotation
import com.mapbox.maps.plugin.annotation.AnnotationConfig
import com.mapbox.maps.plugin.annotation.AnnotationManager
import com.mapbox.maps.plugin.annotation.annotations
import com.mapbox.maps.plugin.annotation.generated.*
import com.mapbox.maps.plugin.annotation.generated.CircleAnnotation
import com.mapbox.maps.plugin.annotation.generated.PolygonAnnotation
import com.mapbox.maps.plugin.annotation.generated.PolylineAnnotation
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.collections.set

class AnnotationController(
  private val mapView: MapView,
  private val messenger: BinaryMessenger,
  private val channelSuffix: String
) : ControllerDelegate {
  private val managerMap = mutableMapOf<String, AnnotationManager<*, *, *, *, *, *, *>>()
  private val streamSinkMap = mutableMapOf<String, InteractionEventsHandler>()
  private val pointAnnotationController = PointAnnotationController(this)
  private val circleAnnotationController = CircleAnnotationController(this)
  private val polygonAnnotationController = PolygonAnnotationController(this)
  private val polylineAnnotationController = PolylineAnnotationController(this)
  private var index = 0

  fun handleCreateManager(call: MethodCall, result: MethodChannel.Result) {
    val id = call.argument<String>("id") ?: (index++).toString()
    val layerId = call.argument<String>("belowLayerId")

    val belowLayerId = if (layerId != null && mapView.mapboxMap.style?.styleLayerExists(layerId) == true) {
      layerId
    } else {
      null
    }
    val manager = when (val type = call.argument<String>("type")!!) {
      "circle" -> {
        mapView.annotations.createCircleAnnotationManager(AnnotationConfig(belowLayerId, id, id)).apply {
          this.addClickListener(
            com.mapbox.maps.plugin.annotation.generated.OnCircleAnnotationClickListener { annotation ->
              sendTapEvent(id, annotation)
              false
            }
          )
          this.addLongClickListener(
            OnCircleAnnotationLongClickListener { annotation ->
              sendLongPressEvent(id, annotation)
              false
            }
          )
          this.addDragListener(object :
              OnCircleAnnotationDragListener {
              override fun onAnnotationDrag(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.CHANGED)
              }

              override fun onAnnotationDragFinished(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.ENDED)
              }

              override fun onAnnotationDragStarted(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.STARTED)
              }
            })
        }
      }
      "point" -> {
        mapView.annotations.createPointAnnotationManager(AnnotationConfig(belowLayerId, id, id)).apply {
          this.addClickListener(
            com.mapbox.maps.plugin.annotation.generated.OnPointAnnotationClickListener { annotation ->
              sendTapEvent(id, annotation)
              false
            }
          )
          this.addLongClickListener(
            OnPointAnnotationLongClickListener { annotation ->
              sendLongPressEvent(id, annotation)
              false
            }
          )
          this.addDragListener(object :
              OnPointAnnotationDragListener {
              override fun onAnnotationDrag(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.CHANGED)
              }

              override fun onAnnotationDragFinished(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.ENDED)
              }

              override fun onAnnotationDragStarted(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.STARTED)
              }
            })
        }
      }
      "polygon" -> {
        mapView.annotations.createPolygonAnnotationManager(AnnotationConfig(belowLayerId, id, id)).apply {
          this.addClickListener(
            com.mapbox.maps.plugin.annotation.generated.OnPolygonAnnotationClickListener { annotation ->
              sendTapEvent(id, annotation)
              false
            }
          )
          this.addLongClickListener(
            OnPolygonAnnotationLongClickListener { annotation ->
              sendLongPressEvent(id, annotation)
              false
            }
          )
          this.addDragListener(object :
              OnPolygonAnnotationDragListener {
              override fun onAnnotationDrag(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.CHANGED)
              }

              override fun onAnnotationDragFinished(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.ENDED)
              }

              override fun onAnnotationDragStarted(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.STARTED)
              }
            })
        }
      }
      "polyline" -> {
        mapView.annotations.createPolylineAnnotationManager(AnnotationConfig(belowLayerId, id, id)).apply {
          this.addClickListener(
            com.mapbox.maps.plugin.annotation.generated.OnPolylineAnnotationClickListener { annotation ->
              sendTapEvent(id, annotation)
              false
            }
          )
          this.addLongClickListener(
            OnPolylineAnnotationLongClickListener { annotation ->
              sendLongPressEvent(id, annotation)
              false
            }
          )
          this.addDragListener(object :
              OnPolylineAnnotationDragListener {
              override fun onAnnotationDrag(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.CHANGED)
              }

              override fun onAnnotationDragFinished(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.ENDED)
              }

              override fun onAnnotationDragStarted(annotation: Annotation<*>) {
                sendDragEvent(id, annotation, GestureState.STARTED)
              }
            })
        }
      }
      else -> {
        result.error("0", "Unrecognized manager type: $type", null)
        return
      }
    }
    managerMap[id] = manager
    val eventsHandler = InteractionEventsHandler()
    eventsHandler.register(
      messenger,
      "$channelSuffix/$id"
    )
    streamSinkMap[id] = eventsHandler
    result.success(id)
  }

  fun handleRemoveManager(call: MethodCall, result: MethodChannel.Result) {
    val id = call.argument<String>("id")!!
    managerMap.remove(id)?.let {
      mapView.annotations.removeAnnotationManager(it)
    }
    result.success(null)
  }

  fun setup() {
    _PointAnnotationMessenger.setUp(messenger, pointAnnotationController, channelSuffix)
    _CircleAnnotationMessenger.setUp(
      messenger,
      circleAnnotationController, channelSuffix
    )
    _PolylineAnnotationMessenger.setUp(
      messenger,
      polylineAnnotationController, channelSuffix
    )
    _PolygonAnnotationMessenger.setUp(
      messenger,
      polygonAnnotationController, channelSuffix
    )
  }

  fun dispose() {
    _PointAnnotationMessenger.setUp(messenger, null, channelSuffix)
    _CircleAnnotationMessenger.setUp(messenger, null, channelSuffix)
    _PolylineAnnotationMessenger.setUp(messenger, null, channelSuffix)
    _PolygonAnnotationMessenger.setUp(messenger, null, channelSuffix)
  }

  fun sendDragEvent(managerId: String, annotation: Annotation<*>, gestureState: GestureState) {
    val context: AnnotationInteractionContext = when (annotation) {
      is com.mapbox.maps.plugin.annotation.generated.PointAnnotation ->
        PointAnnotationInteractionContext(annotation.toFLTPointAnnotation(), gestureState)
      is CircleAnnotation ->
        CircleAnnotationInteractionContext(annotation.toFLTCircleAnnotation(), gestureState)
      is PolygonAnnotation ->
        PolygonAnnotationInteractionContext(annotation.toFLTPolygonAnnotation(), gestureState)
      is PolylineAnnotation ->
        PolylineAnnotationInteractionContext(annotation.toFLTPolylineAnnotation(), gestureState)

      else -> throw IllegalArgumentException("$annotation is unsupported")
    }
    streamSinkMap[managerId]?.dragEventsSink?.success(context)
  }

  fun sendTapEvent(managerId: String, annotation: Annotation<*>) {
    val context: AnnotationInteractionContext = when (annotation) {
      is com.mapbox.maps.plugin.annotation.generated.PointAnnotation ->
        PointAnnotationInteractionContext(annotation.toFLTPointAnnotation(), GestureState.ENDED)
      is CircleAnnotation ->
        CircleAnnotationInteractionContext(annotation.toFLTCircleAnnotation(), GestureState.ENDED)
      is PolygonAnnotation ->
        PolygonAnnotationInteractionContext(annotation.toFLTPolygonAnnotation(), GestureState.ENDED)
      is PolylineAnnotation ->
        PolylineAnnotationInteractionContext(annotation.toFLTPolylineAnnotation(), GestureState.ENDED)

      else -> throw IllegalArgumentException("$annotation is unsupported")
    }
    streamSinkMap[managerId]?.tapEventsSink?.success(context)
  }

  fun sendLongPressEvent(managerId: String, annotation: Annotation<*>) {
    val context: AnnotationInteractionContext = when (annotation) {
      is com.mapbox.maps.plugin.annotation.generated.PointAnnotation ->
        PointAnnotationInteractionContext(annotation.toFLTPointAnnotation(), GestureState.ENDED)
      is CircleAnnotation ->
        CircleAnnotationInteractionContext(annotation.toFLTCircleAnnotation(), GestureState.ENDED)
      is PolygonAnnotation ->
        PolygonAnnotationInteractionContext(annotation.toFLTPolygonAnnotation(), GestureState.ENDED)
      is PolylineAnnotation ->
        PolylineAnnotationInteractionContext(annotation.toFLTPolylineAnnotation(), GestureState.ENDED)

      else -> throw IllegalArgumentException("$annotation is unsupported")
    }
    streamSinkMap[managerId]?.longPressEventsSink?.success(context)
  }

  override fun getManager(managerId: String): AnnotationManager<*, *, *, *, *, *, *> {
    if (managerMap[managerId] == null) {
      throw(Throwable("No manager found with id: $managerId"))
    }
    return managerMap[managerId]!!
  }

  companion object {
    class InteractionEventsHandler {
      var tapEventsSink: PigeonEventSink<AnnotationInteractionContext>? = null
      var dragEventsSink: PigeonEventSink<AnnotationInteractionContext>? = null
      var longPressEventsSink: PigeonEventSink<AnnotationInteractionContext>? = null

      fun register(messenger: BinaryMessenger, channelName: String) {
        val tapEvents = object : AnnotationInteractionEventsStreamHandler() {
          override fun onListen(p0: Any?, sink: PigeonEventSink<AnnotationInteractionContext>) {
            tapEventsSink = sink
          }

          override fun onCancel(p0: Any?) {
            tapEventsSink = null
          }
        }
        val longPressEvents = object : AnnotationInteractionEventsStreamHandler() {
          override fun onListen(p0: Any?, sink: PigeonEventSink<AnnotationInteractionContext>) {
            longPressEventsSink = sink
          }

          override fun onCancel(p0: Any?) {
            longPressEventsSink = null
          }
        }
        val dragEvents = object : AnnotationInteractionEventsStreamHandler() {
          override fun onListen(p0: Any?, sink: PigeonEventSink<AnnotationInteractionContext>) {
            dragEventsSink = sink
          }

          override fun onCancel(p0: Any?) {
            dragEventsSink = null
          }
        }
        AnnotationInteractionEventsStreamHandler.register(messenger, tapEvents, "$channelName/tap")
        AnnotationInteractionEventsStreamHandler.register(messenger, tapEvents, "$channelName/drag")
        AnnotationInteractionEventsStreamHandler.register(messenger, longPressEvents, "$channelName/long_press")
      }
    }
  }
}

File: ios/Classes/AnnotationController.swift
import Foundation
import MapboxMaps
import Flutter

public enum AnnotationControllerError: Error {
    case noManagerFound
    case noAnnotationFound
    case wrongManagerType
}

public protocol ControllerDelegate: AnyObject {
    func getManager(managerId: String) throws -> AnnotationManager
}

extension AnnotationController: AnnotationInteractionDelegate {
    func annotationManager(_ manager: AnnotationManager, didDetectTappedAnnotations annotations: [Annotation]) {
        let annotation = annotations.first
        switch annotation {
        case let annotation as MapboxMaps.PointAnnotation:
            self.onPointAnnotationClickListener?.onPointAnnotationClick(
                annotation: annotation.toFLTPointAnnotation(),
                completion: {_ in }
            )
        case let annotation as MapboxMaps.CircleAnnotation:
            self.onCircleAnnotationClickListener?.onCircleAnnotationClick(
                annotation: annotation.toFLTCircleAnnotation(),
                completion: {_ in }
            )
        case let annotation as MapboxMaps.PolygonAnnotation:
            self.onPolygonAnnotationClickListener?.onPolygonAnnotationClick(
                annotation: annotation.toFLTPolygonAnnotation(),
                completion: {_ in }
            )
        case let annotation as MapboxMaps.PolylineAnnotation:
            self.onPolylineAnnotationClickListener?.onPolylineAnnotationClick(
                annotation: annotation.toFLTPolylineAnnotation(),
                completion: {_ in }
            )
        default:
            print("Can't detemine the type of annotation: \(String(describing: annotation))")
        }
    }
}

class AnnotationController: ControllerDelegate {
    private var mapView: MapView
    private var annotationManagers = [String: AnnotationManager]()
    private var circleAnnotationController: CircleAnnotationController?
    private var pointAnnotationController: PointAnnotationController?
    private var polygonAnnotationController: PolygonAnnotationController?
    private var polylineAnnotationController: PolylineAnnotationController?
    private var onPointAnnotationClickListener: OnPointAnnotationClickListener?
    private var onCircleAnnotationClickListener: OnCircleAnnotationClickListener?
    private var onPolygonAnnotationClickListener: OnPolygonAnnotationClickListener?
    private var onPolylineAnnotationClickListener: OnPolylineAnnotationClickListener?

    init(withMapView mapView: MapView) {
        self.mapView = mapView
        circleAnnotationController = CircleAnnotationController(withDelegate: self)
        pointAnnotationController = PointAnnotationController(withDelegate: self)
        polygonAnnotationController = PolygonAnnotationController(withDelegate: self)
        polylineAnnotationController = PolylineAnnotationController(withDelegate: self)
    }

    func handleCreateManager(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = methodCall.arguments as? [String: Any] else { return }
        guard let type = arguments["type"] as? String else { return }
        let id = (arguments["id"] as? String) ?? String(UUID().uuidString.prefix(5))
        let belowLayerId: String?
        if let layerId = arguments["belowLayerId"] as? String, mapView.mapboxMap.layerExists(withId: layerId) {
            belowLayerId = layerId
        } else {
            belowLayerId = nil
        }

        if let manager = { () -> AnnotationManager? in
            switch type {
            case "circle":
                let circleManager = mapView.annotations.makeCircleAnnotationManager(
                    id: id,
                    layerPosition: belowLayerId.map(MapboxMaps.LayerPosition.below)
                )
                circleManager.delegate = self
                return circleManager
            case "point":
                let pointManager = mapView.annotations.makePointAnnotationManager(
                    id: id,
                    layerPosition: belowLayerId.map(MapboxMaps.LayerPosition.below)
                )
                pointManager.delegate = self
                return pointManager
            case "polygon":
                let polygonManager: PolygonAnnotationManager = mapView.annotations.makePolygonAnnotationManager(
                    id: id,
                    layerPosition: belowLayerId.map(MapboxMaps.LayerPosition.below)
                )
                polygonManager.delegate = self
                return polygonManager
            case "polyline":
                let polylineManager: PolylineAnnotationManager = mapView.annotations.makePolylineAnnotationManager(
                    id: id,
                    layerPosition: belowLayerId.map(MapboxMaps.LayerPosition.below)
                )
                polylineManager.delegate = self
                return polylineManager
            default:
                return nil
            }
        }() {
            annotationManagers[manager.id] = manager
            result(manager.id)
        } else {
            result(AnnotationControllerError.wrongManagerType)
        }
    }

    func handleRemoveManager(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = methodCall.arguments as? [String: Any] else { return }
        guard let id = arguments["id"] as? String else { return }
        annotationManagers.removeValue(forKey: id)
        mapView.annotations.removeAnnotationManager(withId: id)
        result(nil)
    }

    func setup(binaryMessenger: SuffixBinaryMessenger) {
        _CircleAnnotationMessengerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: circleAnnotationController, messageChannelSuffix: binaryMessenger.suffix)
        _PointAnnotationMessengerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: pointAnnotationController, messageChannelSuffix: binaryMessenger.suffix)
        _PolygonAnnotationMessengerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: polygonAnnotationController, messageChannelSuffix: binaryMessenger.suffix)
        _PolylineAnnotationMessengerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: polylineAnnotationController, messageChannelSuffix: binaryMessenger.suffix)
        onPointAnnotationClickListener = OnPointAnnotationClickListener(binaryMessenger: binaryMessenger.messenger, messageChannelSuffix: binaryMessenger.suffix)
        onCircleAnnotationClickListener = OnCircleAnnotationClickListener(binaryMessenger: binaryMessenger.messenger, messageChannelSuffix: binaryMessenger.suffix)
        onPolygonAnnotationClickListener = OnPolygonAnnotationClickListener(binaryMessenger: binaryMessenger.messenger, messageChannelSuffix: binaryMessenger.suffix)
        onPolylineAnnotationClickListener = OnPolylineAnnotationClickListener(binaryMessenger: binaryMessenger.messenger, messageChannelSuffix: binaryMessenger.suffix)
    }

    func tearDown(messenger: SuffixBinaryMessenger) {
        _CircleAnnotationMessengerSetup.setUp(binaryMessenger: messenger.messenger, api: nil, messageChannelSuffix: messenger.suffix)
        _PointAnnotationMessengerSetup.setUp(binaryMessenger: messenger.messenger, api: nil, messageChannelSuffix: messenger.suffix)
        _PolygonAnnotationMessengerSetup.setUp(binaryMessenger: messenger.messenger, api: nil, messageChannelSuffix: messenger.suffix)
        _PolylineAnnotationMessengerSetup.setUp(binaryMessenger: messenger.messenger, api: nil, messageChannelSuffix: messenger.suffix)
        onPointAnnotationClickListener = nil
        onCircleAnnotationClickListener = nil
        onPolygonAnnotationClickListener = nil
        onPolylineAnnotationClickListener = nil
    }

    func getManager(managerId: String) throws -> AnnotationManager {
        if annotationManagers[managerId] == nil {
            throw AnnotationControllerError.noManagerFound
        }
        return annotationManagers[managerId]!
    }
}
